core/src/main/scala/com/mesosphere/usi/core/matching/RangeResource.scala

```diff
@@ -0,0 +1,251 @@
+  * Dynamic values are then selected from the offered ranges. If random implementation is passed in, these ranges are randomized.
+  * For more information about randomization see [[lazyRandomValuesFromRanges()]]
+  *
+  * @param requestedValues values pod wants to consume on the given ranges
+  * @param resourceType name of resource (e.g. ports)
+  * @param random when requesting dynamic values, you can provide Random implementation if you want dynamic values to be randomized
+  */
+case class RangeResource(requestedValues: Seq[Int], resourceType: ResourceType, random: Random = Random)
+    extends ResourceRequirement {
+  override def description: String = s"$resourceType:[${requestedValues.mkString(",")}]"
+
+  override def matchAndConsume(resources: Seq[Protos.Resource]): Option[ResourceMatchResult] = {
+    matchAndConsumeIter(Nil, resources.toList)
+  }
+
+  /**
+    * Iterates over given resources keeping in track unmatched resources.
+    * In the end produces ResourceMatchResult if we were able to find a match on the given resources.
+    *
+    * @param unmatchedResources already processed resources that did not contain resources we are trying to match
+    * @param remainingResources resources we still need to process when looking for resource match
+    * @return None if no match was possible, ResourceMatchResult if we were able to consume requested resources
+    */
+  @tailrec private def matchAndConsumeIter(
+      unmatchedResources: List[Protos.Resource],
+      remainingResources: List[Protos.Resource]): Option[ResourceMatchResult] = {
+
+    remainingResources match {
+      case Nil =>
+        None
+      case next :: rest =>
+        tryConsumeValuesFromResource(requestedValues, next) match {
+          case Nil =>
+            matchAndConsumeIter(next :: unmatchedResources, rest)
+          case consumedResources =>
+            Some(
+              ResourceMatchResult(
+                consumedResources,
+                unmatchedResources ++ rest ++ ResourceUtil.consumeResources(Seq(next), consumedResources)))
+        }
+    }
+  }
+
+  /**
+    * Tries to match and consume values from a given resource
+    * @param values values we want to consume
+    * @param resource mesos resource to match agains
+    * @return final list of mesos resources created after consuming requested values, empty if match was not possible
+    */
+  private def tryConsumeValuesFromResource(values: Seq[Int], resource: Protos.Resource): Seq[Protos.Resource] = {
+    val offeredRanges = parseResourceToRanges(resource)
+    if (offeredRanges.isEmpty || values.isEmpty) {
+      return Seq.empty
+    }
+
+    // non-dynamic values
+    val staticRequestedValues = values.filter(v => v != 0).toSet
+    val availableForDynamicAssignment: Iterator[Int] =
+      lazyRandomValuesFromRanges(offeredRanges, random).filter(v => !staticRequestedValues(v))
```

Comment:
Reuse for what BTW? For multiple pods? I don't see any other reuse here apart from that... And that's not trivial to do right now...

Are you fine with doing such optimizations as a follow up? Could you maybe even create JIRA. I think it would be nice to have some benchmark here as well...

Reply:
`tryConsumeValuesFromResource` is called for each `remainingResources`. We probably should move on but I still don't quite get what's happening here.

Reply:
yes, but for every remaining resource you need to compute new offered ports because there's different range... I don't see any easy optimization here but maybe I am just missing something. I am open to discussing this further...